src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -1,21 +1,464 @@
+  , mciLastMinRfbits        :: !(STRef s (SMaybe Word16))
+    -- | The primary bits of the page-maximum key that we saw last.
+    --
+    -- This should be 'SNothing' if we haven't seen any keys/pages yet.
+  , mciLastMaxPrimbits      :: !(STRef s (SMaybe Word32))
+  }
+
+-- | One-shot construction.
 --
--- * Index lookup (index search)
+fromList :: Int -> [(SerialisedKey, SerialisedKey)] -> CompactIndex
+fromList rfprec ks = runST $ unsafeFreeze =<< fromList' rfprec ks
+
+fromList' :: Int -> [(SerialisedKey, SerialisedKey)] -> ST s (MCompactIndex s)
+fromList' rfprec ks = do
+    mci <- new rfprec n
+    mapM_ (`append` mci) ks
+    pure mci
+  where
+    n = length ks
+
 --
--- * (de-)serialisation
+-- Incremental construction
 --
--- The above list is a sketch. Functionality may move around, and the list is
--- not exhaustive.
+
+{- $incremental
+
+  Constructing a compact index incrementally can go wrong, unless the following
+  preconditions are met and invariants upheld.
+
+  * /Non-empty/: a compact index can not be empty. A mutable compact index can
+    be empty at first, but at least one page should be appended before it is
+    frozen.
+
+  * /Sorted/: pages must be appended in sorted order according to the keys they
+    contain.
+
+  * /Unique/: all keys should be unique within and across pages. A page can
+    contain only one key, in which case the minimum and maximum keys would
+    match.
+
+  * /Partitioned/: pages must be partitioned. That is, the range-finder bits for
+    all keys within a page must match.
+-}
+
+-- | @'new' rfprec n@ creates a new mutable index with a range-finder
+-- bit-precision of @rfprec, and with space reserved for @n@ (disk page)
+-- entries.
 --
-module Database.LSMTree.Internal.Run.Index.Compact () where
+-- Note: after initialisation, the (mutable) index is not resizable, and the
+-- range-finder bit-precision can not be changed.
+new :: Int -> Int -> ST s (MCompactIndex s)
+new rfprec n = MCompactIndex
+    -- Core index structure
+    <$> VUM.new (2 ^ rfprec + 1)
+    <*> pure rfprec
+    <*> VUM.new n
+    <*> VUM.new n
+    <*> newSTRef IntMap.empty
+    -- Aux information required for incremental construction
+    <*> newSTRef 0
+    <*> newSTRef SNothing
+    <*> newSTRef SNothing
+
+-- | Append a new page entry (defined by the minimum and maximum key on the page)
+-- to a mutable compact index.
+append :: (SerialisedKey, SerialisedKey) -> MCompactIndex s -> ST s ()
+append (minKey, maxKey) MCompactIndex{..} = do
+    pageNr <- readSTRef mciCurrentPageNumber
+    writeSTRef mciCurrentPageNumber $ pageNr + 1
+
+    lastMinRfbits <- readSTRef mciLastMinRfbits
+    let lb = smaybe NoBound (\i -> Bound (fromIntegral i) Exclusive) lastMinRfbits
+        ub = Bound (fromIntegral minRfbits) Inclusive
+        x  = fromIntegral pageNr
+    writeRange mciRangeFinder lb ub x
+
+    writeSTRef mciLastMinRfbits (SJust minRfbits)
+
+    VUM.write mciPrimary pageNr minPrimbits
```

Comment:
So presumably this version relies on the bounds checking here to fail if the user `append`s more entries than they had reserved in `new`.